### DISCUSSION TITLE
Run matrix job on macos and on ubuntu in container

### DISCUSSION THREAD

**Discussion by dawidd6** [community/community#26756] [Actions]
<p>Let’s consider the following workflow:</p>
<pre><code>name: On push

on:
  push:
    branches: master

jobs:
  main:
    strategy:
      matrix:
        os:
          - ubuntu-18.04
          - macos-10.14
    runs-on: ${{ matrix.os }}
    container: linuxbrew/brew
    steps:
      - name: brew config
        run: |
          brew config
</code></pre>
<p>Currently the macos job will fail because: “##[error]File not found: ‘docker’”.</p>
<p>What I would want to see is: the macos job would run on macos VM and ubuntu job would run in container.</p>
<p>I know I can workaround this by making 2 separate jobs but I would need to repeat myself and would rather avoid that, granted that this workflow is simple and I can do it easily but it is just an example.</p>

**Comment by ethomson** [comment]
<p>I definitely understand the motivation here.  I’m not sure that just ignoring the container key on macOS runs will necessarily be the solution, but we’ll look at how to simplify this for you so that it’s more DRY.</p>

**Comment by jimhester** [comment]
<p>Just wanted to say this would be very helpful for us as well. Is there anyway currently we can write the yaml so that the container field is included conditionally, only for the linux builds? That would basically solve the issue for us I think.</p>

**Comment by pmatos** [comment]
<p>Yes, the solution here would be for GitHub Actions to allow something like:</p>
<pre><code>container:
  image: debian:stable
  if: matrix.os == ubuntu-latest
</code></pre>
<p>for example.</p>
<p>This problem is compounded by the fact that Yaml anchors are not supported - further forcing us to duplicate CI specification code.</p>

**Comment by aviflax** [comment]
<p><a class="mention" href="/pmatos">@pmatos</a> I like your suggestion! It’s concise, clear, explicit, and consistent. <a class="mention" href="/ethomson">@ethomson</a> what do you think?</p>

**Comment by aguinet** [comment]
<p>For the record, I have exactly the same problem and would really enjoy a solution, like what <a class="mention" href="/pmatos">@pmatos</a> suggests!</p>

**Comment by dawidd6** [comment]
<p><a class="mention" href="/ethomson">@ethomson</a> any update on this?</p>

**Comment by crazy-matt** [comment]
<p>Hi all,<br>
I tried a solution with a matrix job, but ending with a “a step cannot have both the <code>uses</code> and <code>run</code> keys” unfortunately.</p>
<p>The idea was to build this kind of matrix:</p>
<pre><code class="lang-auto">matrix:
  - os: debian
    runnner: ubuntu-latest
    uses: docker://gcr.io/crazy-matt/localenv-builder/debian:latest
  - os: redhat
    runnner: ubuntu-latest
    uses: docker://gcr.io/crazy-matt/localenv-builder/redhat:latest
  - os: suse
    runnner: ubuntu-latest
    uses: docker://gcr.io/crazy-matt/localenv-builder/suse:latest
  - os: mac
    runnner: macos-10.14
    uses: 
</code></pre>
<p>Working on this implementation:</p>
<pre><code class="lang-auto">global_vars:
  runs-on: ubuntu-latest
  outputs:
    matrix: ${{ steps.set_vars.outputs.matrix }}
  steps:
    - uses: actions/checkout@v2
    - id: set_vars
      run: |
        json_array="{\"include\":$(ls -d docker/* | cut -f2 -d'/' | jq --arg repo "docker://ghcr.io/${{ github.repository }}/" --arg tag ":latest" -s -R -c 'split("\n") | to_entries | map( {"os": (.value), "runner": (if .value == "mac" then "macos-10.14" else "ubuntu-latest" end), "uses": (if .value == "mac" then "" else ($repo+.value+$tag) end) })'[:-1])}"
        echo "::set-output name=matrix::$( echo "$json_array" )"

testing:
  runs-on: ${{ matrix.runner }}
  needs: global_vars
  continue-on-error: true
  strategy:
    fail-fast: false
    matrix: ${{ fromJson(needs.global_vars.outputs.matrix) }}
  steps:
    - name: "Running localen-builder for '${{ matrix.os }}'"
      uses: ${{ matrix.uses }}
      run: any_command
</code></pre>
<p>The way you build the matrix json depends on your specific needs.</p>
<p>But I couldn’t make it work.<br>
Any news on this <a class="mention" href="/ethomson">@ethomson</a></p>

**Comment by HannesHaglund** [comment]
<p>I have the same problem.</p>
<p><a class="mention" href="/ethomson">@ethomson</a> Any update?</p>
<p>This practically not being supported forces anyone who needs to support executing items both in and outside containers to copy and paste all of their jobs, which is silly.</p>
<p>Either adding support for an if keyword to the container keyword or adding support for Yaml Anchors, as <a class="mention" href="/pmatos">@pmatos</a> wrote, would solve this issue for me.</p>
<p>Solving <a href="https://github.com/actions/runner/issues/646" rel="noopener nofollow ugc">https://github.com/actions/runner/issues/646</a> might also help. At least you wont have to copy all of your steps from job to job, as you could wrap all of your steps inside a composite action, but that feels like a bit of a band-aid.</p>
<p>Thank you.</p>

**Comment by dawidd6** [comment, marked as answer]
<aside class="onebox githubissue" data-onebox-src="https://github.com/actions/runner/issues/265">
  <header class="source">

      <a href="https://github.com/actions/runner/issues/265" target="_blank" rel="noopener nofollow ugc">github.com/actions/runner</a>
  </header>

  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Issue">
	  <svg width="60" height="60" class="github-icon" viewbox="0 0 14 16" aria-hidden="true"><path d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7s-2.56 5.7-5.7 5.7A5.71 5.71 0 0 1 1.3 8c0-3.14 2.56-5.7 5.7-5.7zM7 1C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7-3.14-7-7-7zm1 3H6v5h2V4zm0 6H6v2h2v-2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/actions/runner/issues/265" target="_blank" rel="noopener nofollow ugc">Allow an empty value to be specified for jobs.&lt;id&gt;.container</a>
    </h4>

    <div class="github-info">
      <div class="date">
        opened <span class="discourse-local-date" data-format="ll" data-date="2020-01-12" data-time="03:32:06" data-timezone="UTC">03:32AM - 12 Jan 20 UTC</span>
      </div>

        <div class="date">
          closed <span class="discourse-local-date" data-format="ll" data-date="2020-01-13" data-time="04:02:37" data-timezone="UTC">04:02AM - 13 Jan 20 UTC</span>
        </div>

      <div class="user">
        <a href="https://github.com/ericsciple" target="_blank" rel="noopener nofollow ugc">
          <img alt="ericsciple" src="https://user-images.githubusercontent.com/9713907/181099779-85814ce4-9fa6-4c57-865c-f520f97e96ea.jpeg" class="onebox-avatar-inline" width="20" height="20">
          ericsciple
        </a>
      </div>
    </div>

    <div class="labels">
        <span style="display:inline-block;margin-top:2px;background-color: #B8B8B8;padding: 2px;border-radius: 4px;color: #fff;margin-left: 3px;">
          enhancement
        </span>
    </div>
  </div>
</div>

  <div class="github-row">
    <p class="github-body-container">**Is your feature request related to a problem? Please describe.**

I'd like t<span class="show-more-container"><a href="" rel="noopener" class="show-more">…</a></span><span class="excerpt hidden">o be able to build a matrix workflow where some jobs run in a container and some run directly on the VM.  I may have a cross-platform tool where I have strange dependencies that I provide via a container, or I want to run on multiple different Linux distributions, but also run on Windows and macOS.  So I would like a way to specify that I run on _no_ container.  This would allow me to create a workflow like:

```yaml
name: CI

on: [push]

jobs:    
  build:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        image: ['ubuntu:latest', 'centos:latest', '']
```

A more realistic scenario would be a matrix of `runs-on` that includes linux, windows and macos, and then `include`ing the container value on linux hosts.

**Describe the solution you'd like**
That some value exists for `container` that means "none".  Perhaps: an empty string (`''`), a dash (`'-'`), or a dot (`'.'`), all of which are not legal image names.</span></p>
  </div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>
